streams/src/main/java/org/apache/kafka/streams/kstream/Branched.java

```diff
@@ -0,0 +1,150 @@
+                       final Consumer<? super KStream<K, V>> chainConsumer) {
+        this.name = name;
+        this.chainFunction = chainFunction;
+        this.chainConsumer = chainConsumer;
+    }
+
+    /**
+     * Create an instance of {@link Branched} from an existing instance.
+     *
+     * @param branched the instance of {@link Branched} to copy
+     */
+    protected Branched(final Branched<K, V> branched) {
+        this(branched.name, branched.chainFunction, branched.chainConsumer);
+    }
+
+    /**
+     * Configure the instance of {@link Branched} with a branch name postfix.
+     *
+     * @param name the branch name postfix to be used. If {@code null} a default branch name postfix will be generated (see
+     *             {@link BranchedKStream} description for details)
+     * @return {@code this}
+     */
+    @Override
+    public Branched<K, V> withName(final String name) {
+        return new Branched<>(name, chainFunction, chainConsumer);
+    }
+
+    /**
+     * Create an instance of {@link Branched} with provided branch name postfix.
+     *
+     * @param name the branch name postfix to be used. If {@code null}, a default branch name postfix will be generated
+     *             (see {@link BranchedKStream} description for details)
+     * @param <K>  key type
+     * @param <V>  value type
+     * @return a new instance of {@link Branched}
+     */
+    public static <K, V> Branched<K, V> as(final String name) {
+        return new Branched<>(name, null, null);
+    }
+
+    /**
+     * Create an instance of {@link Branched} with provided chain function.
+     *
+     * @param chain A function that will be applied to the branch. If {@code null}, the identity
+     *              {@code kStream -> kStream} function will be supposed. If this function returns
+     *              {@code null}, its result is ignored, otherwise it is added to the {@code Map} returned
+     *              by {@link BranchedKStream#defaultBranch()} or {@link BranchedKStream#noDefaultBranch()} (see
+     *              {@link BranchedKStream} description for details).
+     * @param <K>   key type
+     * @param <V>   value type
+     * @return a new instance of {@link Branched}
+     */
+    public static <K, V> Branched<K, V> withFunction(
+            final Function<? super KStream<K, V>,
+                    ? extends KStream<K, V>> chain) {
+        return new Branched<>(null, chain, null);
+    }
+
+    /**
+     * Create an instance of {@link Branched} with provided chain consumer.
+     *
+     * @param chain A consumer to which the branch will be sent. If a non-null branch is provided here,
+     *              the respective branch will not be added to the resulting {@code Map} returned
+     *              by {@link BranchedKStream#defaultBranch()} or {@link BranchedKStream#noDefaultBranch()} (see
+     *              {@link BranchedKStream} description for details). If {@code null}, a no-op consumer will be supposed

```

Comment:
OK let me explain how I understood this: 

Let's try to forget about Java typing system and look at the problem conceptually, from set / category theory point of view

We're using either **function** or **consumer** or **nothing** for branch processing. From mathematical point of view, we can consider consumer to be a function that maps its domain to empty set. This explains our decision why consumer 'swallows' branch unlike function that 'forwards' its result to the resulting map. When we provide **nothing**, like in this example

```java
BranchedKStream<Integer, String> branch = source.split().branch(isEven).branch(isMultipleOfSeven);
```
we agreed to treat this like 

```java
        BranchedKStream<Integer, String> branch = source.split()
                .branch(isEven, Branched.withFunction(Function.identity()))
                .branch(isMultipleOfSeven, Branched.withFunction(Function.identity()));
```

Now back to Java: in Java, for `nothing` we have `null`. My original intention was to try to emulate `Function|Consumer` union type using overloading, but Java didn't allow me this dirty trick -- thus `withFunction` and `withConsumer` builder methods appeared 😄 

If I had succeded, it would have made sence to treat `null` uniformely for both functions and consumers. 

Hope this explains why I allowed `null`s and always treated it as an identity function.

But now I think you're right. I'm ready to just prohibit passing in nulls and throw explicit NPEs for both functions and consumers. This, at least, can spare us from further argumets :-) What do you think?

Reply:
SGTM. In general, even if possible, I don't like to give `null` semantics :)